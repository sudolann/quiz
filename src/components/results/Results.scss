@import '../../styles/constants.scss';
@import '../../styles/constants.scss';

.results {
    max-width: 700px;
    margin: 0 auto;
    min-width: 300px;
    padding: 1rem 3rem;

    #quiz-btn {
        @media (min-width: $breakpoint-desktop) {
            padding: 0.8rem 1.95rem;
        }
    }
}

.results__title {
    margin-bottom: 1.5rem;
}

.results__questions {
    display: flex;
    flex-direction: column;
    align-items: left;
    justify-content: center;

    #quiz-btn {
        align-self: center;
    }
}

.results__question__item {
    display: flex;
    justify-content: left;
    align-items: center;
    width: 100%;
    border-bottom: 1px solid $light-grey;
    padding: 0.7rem 0;


    @media (min-width: $breakpoint-tablet) {
        padding: 1rem 0;
    }

    @media (min-width: $breakpoint-desktop) {
        padding: 1.2rem 0;
    }

    &:last-of-type {
        margin-bottom: 1rem;
    }

    .svg {
        padding: 0.4rem;
    }

    svg {
        width: 0.8rem;
        height: auto;

        @media (min-width: $breakpoint-tablet) {
            width: 1rem;
        }

        @media (min-width: $breakpoint-desktop) {
            min-height: 1.2rem;
        }
    }

}