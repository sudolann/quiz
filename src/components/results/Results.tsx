import React, { FunctionComponent, ReactElement, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { QuestionsProps } from '../../state/quiz/reducer';
import { Button } from 'antd';
import { ReactComponent as Minus } from '../../assets/minus.svg';
import { ReactComponent as Plus } from '../../assets/plus.svg';
import { resetQuiz } from '../../state/quiz/actions';
import { resetScore } from '../../state/points/actions';
import { useHistory } from 'react-router';
import { StateProps, store } from '../../state';
import './Results.scss';

export const Results: FunctionComponent = (): ReactElement => {
  const dispatch = useDispatch();
  const history = useHistory();
  const state = useSelector<ReturnType<typeof store.getState>, StateProps>((state) => state);

  const points: Array<0 | 1> = state.points;
  const list: QuestionsProps[] = state.quiz.list;

  const sum = points.reduce((acc: number, item: number) => {
    const res = acc + item;
    return res;
  }, 0);
  useEffect(() => {
    if (!list.length || !points.length) {
      history.push('/');
    }
  }, [points, list]);

  const handleClick = () => {
    dispatch(resetQuiz());
    dispatch(resetScore());
  };
  return (
    <div className='results'>
      <div className='results__title'>
        <h1>
          You scored
          <br />
          {sum}/{points.length}
        </h1>
      </div>
      <div className='results__questions'>
        {list.map((item: QuestionsProps, index: number) => (
          <div className='results__question__item' key={index}>
            <div className='svg'>{points[index] === 1 ? <Plus /> : <Minus />}</div>
            <p>{item.question}</p>
          </div>
        ))}
        <Button type='primary' id='quiz-btn' onClick={handleClick}>
          play again?
        </Button>
      </div>
    </div>
  );
};
