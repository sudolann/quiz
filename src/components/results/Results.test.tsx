import { shallow } from 'enzyme';
import React from 'react';
import { Results } from './Results';
import renderer from 'react-test-renderer';
import * as redux from 'react-redux';
import { ReactComponent as Minus } from '../../assets/minus.svg';
import { ReactComponent as Plus } from '../../assets/plus.svg';

const mock = jest.fn();

const spy = jest.spyOn(redux, 'useSelector');
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: (): jest.Mock<any, any> => mock,
}));
jest.mock('react-router', () => ({
  useHistory: () => ({
    push: mock,
  }),
}));
const quiz = {
  list: [
    {
      category: 'General Knowledge',
      question: 'Nutella is produced by the German company Ferrero.',
      correct_answer: 'False',
    },
    { category: 'General Knowledge', question: 'Romanian belongs to the Romance language family, shared with French, Spanish, Portuguese and Italian. ', correct_answer: 'True' },
  ],
  loading: false,
};

const points = [0, 1];

describe('component', (): void => {
  describe('Results', (): void => {
    beforeEach(() => {
      spy.mockReturnValue({
        quiz,
        points,
      });
    });
    it('match snapshot', (): void => {
      const tree = renderer.create(<Results />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders questions with proper svg icon', (): void => {});
    spy.mockReturnValue({
      quiz,
      points,
    });
    const wrapper = shallow(<Results />);
    wrapper.find('.results__question__item').map((item, index) => {
      expect(item.find('p').text()).toBe(quiz.list[index].question);
      const svg = item.find('.svg');

      if (points[index] === 0) {
        expect(svg.find(Minus)).toHaveLength(1);
      }

      if (points[index] === 1) {
        expect(svg.find(Plus)).toHaveLength(1);
      }
    });
  });

  it('calls dispatch and history upon clicking button', (): void => {});
  spy.mockReturnValue({
    quiz,
    points,
  });
  const wrapper = shallow(<Results />);
  const button = wrapper.find('Button');

  button.simulate('click');
  expect(mock).toHaveBeenCalledTimes(3);
});
