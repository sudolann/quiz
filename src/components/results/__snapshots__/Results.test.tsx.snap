// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Results match snapshot 1`] = `
<div
  className="results"
>
  <div
    className="results__title"
  >
    <h1>
      You scored
      <br />
      1
      /
      2
    </h1>
  </div>
  <div
    className="results__questions"
  >
    <div
      className="results__question__item"
    >
      <div
        className="svg"
      >
        <svg>
          minus.svg
        </svg>
      </div>
      <p>
        Nutella is produced by the German company Ferrero.
      </p>
    </div>
    <div
      className="results__question__item"
    >
      <div
        className="svg"
      >
        <svg>
          plus.svg
        </svg>
      </div>
      <p>
        Romanian belongs to the Romance language family, shared with French, Spanish, Portuguese and Italian. 
      </p>
    </div>
    <button
      className="ant-btn ant-btn-primary"
      id="quiz-btn"
      onClick={[Function]}
      type="button"
    >
      <span>
        play again?
      </span>
    </button>
  </div>
</div>
`;
